### PR TITLE
PyTorchでFarmerを動かせるようにする

### PR DESCRIPTION
**Pull request by curogihu**
なお、クロスエントロピーに関するバグ解決済み。

1. data.Datasetでワンホットコーディングしてない、素のラベルを戻り値に追加
2. モデルの学習でクロスエントロピーを行う際、手順1で返した素のラベルを使用する

### CODE REVIEW

File: farmer/domain/tasks/evaluation_task.py [hunk 1]
```diff
@@ -14,31 +14,56 @@ def command(self, annotation_set, model=None, prediction=None):
...
-                    "--convert-model"
-                ]
-            )
-            eval_report = ["evaluation result is std out"]
-        return eval_report
+                    self.config.train_colors
+                )
+
+            elif self.config.framework == "pytorch":
+                return ncc.metrics.iou_dice_val_pytorch(
```
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/evaluation_task.py]: PyTorch独自の処理が途中でいくつも入り、そのために分岐処理を書くことで
メンテナンス性が落ちる為、あえて分けました。

なお、Tensorflow, PyTorchでの処理の違いは以下の通りです。

Tensorflow
・すでにGPU設定済み
・モデルの推論はmode.predict_on_batchで済む

PyTorch: 
・モデルを推論モードに設定する
・PyTorchでの推論の為に次元を並び替える
・推論対象の画像にGPUを設定する
・Tensorflowで実施する推論後の処理を流用する為、推論結果の次元を並び替える
・推論結果をNumpyに変換する

File: farmer/domain/tasks/build_model_task.py [hunk 2]
```diff
@@ -10,6 +12,10 @@
 
 from tensorflow import keras
 
+import torch
+import torch.nn as nn
+import torch.optim as optim
```
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/build_model_task.py]: import torch.nn as nnではなく、import torch.nn as torch.nnのようにした方がいいのではないかということですが、そう思いません。

一番の理由は、慣習的に使用されている別名が
２つのフレームワークで、バッティングしない為です。
例：keras.*, torch.nn as nn, torch.optim as optim

また、tensorflowに内蔵されているkerasを使うので
from tensorflow import kerasと記載するのであり、
torch.hogeなどと無理に合わせる必要はないと思います。

File: farmer/domain/tasks/build_model_task.py [hunk 3]
```diff
@@ -1,4 +1,6 @@
 import segmentation_models
+import segmentation_models_pytorch as smp
+
 from segmentation_models import Unet, PSPNet
```
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/build_model_task.py]: ごもっともなので、後ほど対応致します。

File: farmer/domain/tasks/build_model_task.py [hunk 4]
```diff
@@ -61,90 +70,144 @@ def _do_make_model_task(
...
-                model = Model2D(nb_classes, height, width)
+                if model_name == "xception" and xception_shape_condition:
+                    model = xception(nb_classes, height, width)
+                elif model_name == "mobilenet" and mobilenet_shape_condition:
+                    model = mobilenet(nb_classes, height, width)
+                else:
+                    model = Model2D(nb_classes, height, width)
+
+            elif self.config.framework == 'pytorch':
+                model = None
```
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/build_model_task.py]: ごもっともなので、後ほどコメントを残しておきます。

File: farmer/domain/tasks/evaluation_task.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yumion) [reply farmer/domain/tasks/evaluation_task.py]: じゃあ別関数として元に戻しておきます

File: farmer/domain/tasks/train_task.py [hunk 5]
```diff
@@ -3,28 +3,53 @@
 from farmer import ncc
 from tensorflow.python import keras
```
Comment (yumion) [inline farmer/domain/tasks/train_task.py]: @aiorhiroki `from tensorflow.python import keras` とimport してる場合と， `from tensorflow import keras` としてる場合がありますが，どっちかに統一した方がよくないですか？
(前のfarmerでtf.kerasにした時は`from tensorflow.python import keras` にしていたと思います．)

File: farmer/domain/tasks/build_model_task.py [hunk 6]
```diff
@@ -61,90 +70,144 @@ def _do_make_model_task(
...
+                        in_channels=3,
+                        classes=nb_classes,
+                    )
+
+                elif model_name == "pspnet":
+                    model = smp.PSPNet(
+                        encoder_name=backbone,
+                        in_channels=3,
+                        classes=nb_classes,
+                    )
```
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/build_model_task.py]: Tensorflowの場合 ※以下のURLより抜粋
https://github.com/qubvel/segmentation_models/blob/master/segmentation_models/models/unet.py
```
def Unet(
        backbone_name='vgg16',
        input_shape=(None, None, 3),
        classes=1,
        activation='sigmoid',
        weights=None,
        encoder_weights='imagenet',
        encoder_freeze=False,
        encoder_features='default',
        decoder_block_type='upsampling',
        decoder_filters=(256, 128, 64, 32, 16),
        decoder_use_batchnorm=True,
        **kwargs
):
```
PyTorchの場合　※以下のURLより抜粋
https://github.com/qubvel/segmentation_models.pytorch/blob/master/segmentation_models_pytorch/unet/model.py
```
class Unet(SegmentationModel):
def __init__(
        self,
        encoder_name: str = "resnet34",
        encoder_depth: int = 5,
        encoder_weights: str = "imagenet",
        decoder_use_batchnorm: bool = True,
        decoder_channels: List[int] = (256, 128, 64, 32, 16),
        decoder_attention_type: Optional[str] = None,
        in_channels: int = 3,
        classes: int = 1,
        activation: Optional[Union[str, callable]] = None,
        aux_params: Optional[dict] = None,
    ):
```

File: farmer/domain/tasks/build_model_task.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yumion) [reply farmer/domain/tasks/build_model_task.py]: そうですか．kerasも`from tensorflow.keras import optimizer` とかやると思いますが，やってないのはkerasの関数と明確にするためだからtorchも合わせた方がスパゲティコードでも読みやすいのかなと思ったのですが．

File: farmer/domain/tasks/build_model_task.py [hunk 7]
```diff
@@ -61,90 +70,144 @@ def _do_make_model_task(
...
+                    model = Deeplabv3(
+                        input_shape=(height, width, 3),
+                        classes=nb_classes,
+                        backbone=backbone,
+                    )
+
+                elif model_name == "pspnet":
+                    model = PSPNet(
+                        backbone_name=backbone,
+                        input_shape=(height, width, 3),
```
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/build_model_task.py]: Tensorflow部分は担当外なので、ノータッチの予定。
すごく心配であれば、松崎さんに連絡し、対応方法を仰ぐ。

File: farmer/domain/tasks/predict_classification_task.py [hunk 8]
```diff
@@ -16,42 +16,80 @@ def command(self, test_set, model, save_npy=True):
...
+                steps=len(annotation_gen),
+                workers=16 if self.config.multi_gpu else 1,
+                max_queue_size=32 if self.config.multi_gpu else 10,
+                use_multiprocessing=self.config.multi_gpu,
+                verbose=1,
+            )
+
+        elif self.config.framework == "pytorch":
+            print(
+                "[_do_classification_predict_task, pytorch] it is under construction.")
```
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/predict_classification_task.py]: 実装すべき処理をコメントで残しておきます。

NotImplementedErrorはスーパークラスを継承したサブクラスで、
オーバーライド漏れの関数があったときのエラーなので、不適格と思われます。

File: farmer/domain/tasks/predict_classification_task.py [hunk 9]
```diff
@@ -16,42 +16,80 @@ def command(self, test_set, model, save_npy=True):
...
+                    [
+                        image_file,
+                        self.config.class_names[int(prediction_cls)]
+                    ]
+                )
+            with open(f"{self.config.info_path}/pred.csv", "w") as fw:
+                writer = csv.writer(fw)
+                writer.writerows(pred_result)
+
+        elif self.config.framework == "pytorch":
```
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/predict_classification_task.py]: PyTorchでも同じコードで動作するか確認
→ 問題がなければ、完全一本化

File: farmer/ncc/utils/image.py [hunk 10]
```diff
@@ -106,6 +110,49 @@ def generate_segmentation_result(
         result_image.save(f"{save_dir}/{save_image_name}")
 
 
+def generate_segmentation_result_pytorch(
```
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/ncc/utils/image.py]: 引数にframeworkを入れ、処理の途中で
framework毎の条件分岐を入れれば、一本化は可能です。

しかし、行う場合はリファクタリングといえども、
コードを修正したTensorflowも正常動作するかのテストが発生します。

テストの労力に見合った成果は期待できない為、分けました。

File: farmer/domain/tasks/predict_classification_task.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yumion) [reply farmer/domain/tasks/predict_classification_task.py]: コメントアウトはわかりますがprintでコメントは普通しなくないですか？
`build_mode_task.py` の `def _do_make_model_task` でもNotImplementedErrorを使っていたので

File: farmer/domain/tasks/predict_classification_task.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/predict_classification_task.py]: Tensorflow独自の関数がなさそうなので、
Tensorflowのコードをコピペしてます。

Classification実装時に試行して、
流用大丈夫そうなら一本化で。

File: farmer/domain/tasks/train_task.py [hunk 11]
```diff
@@ -117,28 +278,264 @@ def _do_set_callbacks_task(self, base_model, train_set, validation_set):
...
-            return base_model
+    def _do_fetch_criterion(self, criterion_name: str):
+        if criterion_name == 'dice_loss':
+            return DiceLoss()
+
+        elif criterion_name == 'jaccard_loss':
+            return JaccardLoss()
+
+        elif criterion_name == 'crossentropy_loss':
+            # 一旦、ダイスロスにしておく
```
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/train_task.py]: コメントアウトの件、了解です。
CrossEntropyの件は解決済みです。

以下、詳細

```
from segmentation_models_pytorch.utils.losses import (
    DiceLoss, JaccardLoss, CrossEntropyLoss
)

def _do_fetch_criterion(self, criterion_name: str):
    if criterion_name == 'dice_loss':
        return DiceLoss()

    elif criterion_name == 'jaccard_loss':
        return JaccardLoss()

    elif criterion_name == 'crossentropy_loss':
        return CrossEntropyLoss()

    else:
        return None

```

File: farmer/domain/tasks/train_task.py [hunk 12]
```diff
@@ -117,28 +278,264 @@ def _do_set_callbacks_task(self, base_model, train_set, validation_set):
...
+            if self.config.multi_gpu:
+                torch.save(base_model.state_dict(), model_path)
+                return base_model
+
+            else:
+                torch.save(model.state_dict(), model_path)
+                return model
+
+
+class CrossEntropy(nn.Module):
```
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/train_task.py]: 消してOKです。

以下で対応しています。
```
from segmentation_models_pytorch.utils.losses import (
    DiceLoss, JaccardLoss, CrossEntropyLoss
)
```

File: farmer/domain/tasks/set_train_env_task.py [hunk 13]
```diff
@@ -37,6 +43,9 @@ def _do_set_cpu_gpu_devices_task(self, gpu: str):
             tf.config.threading.set_inter_op_parallelism_threads(num_threads)
             tf.config.threading.set_intra_op_parallelism_threads(num_threads)
 
+        # elif self.config.framework == 'pytorch':
```
[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/set_train_env_task.py]: domain/tasks/set_train_env_task.pyで
seed固定の漏れがあったので、以下のように対応します。

詳細は下記参考
torch.cuda.manual_seedとtorch.cuda.manual_seed_allの違い
※torch.cuda.manual_seed(seed)のwarningを参照
https://pytorch.org/docs/stable/cuda.html

```
def _do_set_random_seed_task(self, seed=1):
        # set random_seed
        os.environ["PYTHONHASHSEED"] = str(seed)
        np.random.seed(seed)
        rn.seed(seed)
        if self.config.framework == "tensorflow":
            tf.random.set_seed(seed)
            # tf.set_random_seed(seed)

        elif self.config.framework == 'pytorch':
            torch.manual_seed(seed)
            torch.cuda.manual_seed(seed)　　　　      # for single GPU
            torch.cuda.manual_seed_all(seed)         # for multiple GPU

            torch.backends.cudnn.benchmark = False
            torch.backends.cudnn.deterministic = True
```

[the remark this answers is not in the document]
Comment (curogihu) [reply farmer/domain/tasks/set_train_env_task.py]: 以下画像は、





seed固定の確認の件

File: farmer/domain/model/config_model.py [hunk 14]
```diff
@@ -19,4 +19,5 @@ class Config:
     learning_path: str = None
     image_path: str = None
     return_result: bool = False
-    framework: str = "tensorflow"
+    # framework: str = "tensorflow"
```
Comment (aiorhiroki) [inline farmer/domain/model/config_model.py:22]: コメントアウト削除

File: farmer/domain/model/task_model.py [hunk 15]
```diff
@@ -1,7 +1,12 @@
...
 
 class Task(IntEnum):
     CLASSIFICATION = 1
     OBJECT_DETECTION = 2
     SEMANTIC_SEGMENTATION = 3
+
+
+class Framework(Enum):
+    TENSORFLOW = "tensorflow"
+    PYTORCH = "pytorch"
```
Comment (aiorhiroki) [inline farmer/domain/model/task_model.py:12]: これ良いのに、使ってなくないか？
あと、Taskみたいに、IntEnumが良さそう
